strawberry/asgi/__init__.py

```diff
@@ -59,38 +59,70 @@ async def get_context(
         return {"request": request}
 
     async def handle_keep_alive(self, websocket):
-        if websocket.application_state == WebSocketState.DISCONNECTED:
+        if (
+            websocket.application_state == WebSocketState.DISCONNECTED
+        ):  # pragma: no cover
             return
 
-        await websocket.send_json({"type": GQL_CONNECTION_KEEP_ALIVE})
         await asyncio.sleep(self.keep_alive_interval)
+        await websocket.send_json({"type": GQL_CONNECTION_KEEP_ALIVE})
 
         self._keep_alive_task = asyncio.create_task(self.handle_keep_alive(websocket))
 
     async def handle_websocket(self, scope: Scope, receive: Receive, send: Send):
         websocket = WebSocket(scope=scope, receive=receive, send=send)
 
-        await websocket.accept(subprotocol="graphql-ws")
-
-        while websocket.application_state != WebSocketState.DISCONNECTED:
-            message = await websocket.receive_json()
-
-            operation_id = message.get("id")
-            message_type = message.get("type")
+        subscriptions: typing.Dict[str, typing.AsyncGenerator] = {}
+        tasks = {}
 
-            if message_type == GQL_CONNECTION_INIT:
-                await websocket.send_json({"type": GQL_CONNECTION_ACK})
+        await websocket.accept(subprotocol="graphql-ws")
 
-                if self.keep_alive:
-                    asyncio.create_task(self.handle_keep_alive(websocket))
-            elif message_type == GQL_CONNECTION_TERMINATE:
-                await websocket.close()
-            elif message_type == GQL_START:
-                await self.start_subscription(
-                    message.get("payload"), operation_id, websocket
-                )
-            elif message_type == GQL_STOP:
-                await websocket.close()
+        try:
```

Comment:
Any chance this `try/catch` could be refactored a bit? It's quite long and daunting to look at? I think a good start would be to move each `message_type` switch/case to a dedicated function?